File: rxandroid/src/main/java/rx/android/widget/RxRatingBar.java
package rx.android.widget;

import android.widget.RatingBar;
import rx.Observable;
import rx.functions.Action1;

public final class RxRatingBar {
  /**
   * Create an observable of the rating changes on {@code view}.
   * <p>
   * <em>Warning:</em> The created observable keeps a strong reference to {@code view}. Unsubscribe
   * to free this reference.
   */
  public static Observable<Float> ratingChanges(RatingBar view) {
    return Observable.create(new RatingBarRatingChangeOnSubscribe(view));
  }

  /**
   * Create an observable of the rating change events on {@code view}.
   * <p>
   * <em>Warning:</em> The created observable keeps a strong reference to {@code view}. Unsubscribe
   * to free this reference.
   */
  public static Observable<RatingBarChangeEvent> ratingChangeEvents(RatingBar view) {
    return Observable.create(new RatingBarRatingChangeEventOnSubscribe(view));
  }

  /**
   * An action which sets the rating of {@code view}.
   * <p>
   * <em>Warning:</em> The created observable keeps a strong reference to {@code view}. Unsubscribe
   * to free this reference.
   */
  public static Action1<? super Float> setRating(final RatingBar view) {
    return new Action1<Float>() {
      @Override public void call(Float value) {
        view.setRating(value);
      }
    };
  }

  /**
   * An action which sets whether {@code view} is an indicator (thus non-changeable by the user).
   * <p>
   * <em>Warning:</em> The created observable keeps a strong reference to {@code view}. Unsubscribe
   * to free this reference.
   */
  public static Action1<? super Boolean> setIsIndicator(final RatingBar view) {
    return new Action1<Boolean>() {
      @Override public void call(Boolean value) {
        view.setIsIndicator(value);
      }
    };
  }
}

File: rxandroid-kotlin/src/main/kotlin/rx/android/widget/RxRatingBar.kt
package rx.android.widget

import android.widget.RatingBar
import rx.Observable
import rx.functions.Action1

/**
 * Create an observable of the rating changes on {@code view}.
 * <p>
 * <em>Warning:</em> The created observable keeps a strong reference to {@code view}. Unsubscribe
 * to free this reference.
 */
public inline fun RatingBar.ratingChanges(): Observable<Float> = RxRatingBar.ratingChanges(this)

/**
 * Create an observable of the rating change events on {@code view}.
 * <p>
 * <em>Warning:</em> The created observable keeps a strong reference to {@code view}. Unsubscribe
 * to free this reference.
 */
public inline fun RatingBar.ratingChangeEvents(): Observable<RatingBarChangeEvent> = RxRatingBar.ratingChangeEvents(this)

/**
 * An action which sets the rating of {@code view}.
 * <p>
 * <em>Warning:</em> The created observable keeps a strong reference to {@code view}. Unsubscribe
 * to free this reference.
 */
public inline fun RatingBar.rating(): Action1<in Float> = RxRatingBar.setRating(this)

/**
 * An action which sets whether {@code view} is an indicator (thus non-changeable by the user).
 * <p>
 * <em>Warning:</em> The created observable keeps a strong reference to {@code view}. Unsubscribe
 * to free this reference.
 */
public inline fun RatingBar.isIndicator(): Action1<in Boolean> = RxRatingBar.setIsIndicator(this)
